### ISSUE TITLE
Compilation in pedantic mode produces errors (redefinition of typedef)

### ISSUE THREAD

**Issue by inobelar** [warmcat/libwebsockets#3002]
During compilation of my own library (compiled with `C99` standard & `-pedantic -pedantic-errors` flags), that internally uses `libwebsockets` (version: currently latest commit [65f977c](https://github.com/warmcat/libwebsockets/tree/65f977c7dee092662045c84662b403960c0f3e86) in `main` branch), I found the next compilation errors:

## Redefinition of typedef `lws_display_state_t` 

```
In file included from /path/to/libwebsockets_installed/include/libwebsockets.h:786,
                 from /path/to/my_lib/source.c:
/path/to/libwebsockets_installed/include/libwebsockets/lws-dlo.h:270:34: error: redefinition of typedef ‘lws_display_state_t’ [-Wpedantic]
  270 | typedef struct lws_display_state lws_display_state_t;
      |                                  ^~~~~~~~~~~~~~~~~~~

In file included from /path/to/libwebsockets_installed/include/libwebsockets.h:785,
                 from /path/to/my_lib/source.c:
/path/to/libwebsockets_installed/include/libwebsockets/lws-display.h:162:3: note: previous declaration of ‘lws_display_state_t’ with type ‘lws_display_state_t’ {aka ‘struct lws_display_state’}
  162 | } lws_display_state_t;
      |   ^~~~~~~~~~~~~~~~~~~
```

References:
https://github.com/warmcat/libwebsockets/blob/65f977c7dee092662045c84662b403960c0f3e86/include/libwebsockets/lws-display.h#L139-L162
https://github.com/warmcat/libwebsockets/blob/65f977c7dee092662045c84662b403960c0f3e86/include/libwebsockets/lws-dlo.h#L270

## Redefinition of typedef `lhp_ctx_t`

```
In file included from /path/to/libwebsockets_installed/include/libwebsockets.h:797,
                 from /path/to/my_lib/source.c:
/path/to/libwebsockets_installed/include/libwebsockets/lws-html.h:588:3: error: redefinition of typedef ‘lhp_ctx_t’ [-Wpedantic]
  588 | } lhp_ctx_t;
      |   ^~~~~~~~~

In file included from /path/to/libwebsockets_installed/include/libwebsockets.h:786,
                 from /path/to/my_lib/source.c:
/path/to/libwebsockets_installed/include/libwebsockets/lws-dlo.h:463:24: note: previous declaration of ‘lhp_ctx_t’ with type ‘lhp_ctx_t’ {aka ‘struct lhp_ctx’}
  463 | typedef struct lhp_ctx lhp_ctx_t;
      |                        ^~~~~~~~~
```

References:
https://github.com/warmcat/libwebsockets/blob/65f977c7dee092662045c84662b403960c0f3e86/include/libwebsockets/lws-html.h#L512-L588
https://github.com/warmcat/libwebsockets/blob/65f977c7dee092662045c84662b403960c0f3e86/include/libwebsockets/lws-dlo.h#L463

----

## Additional info
- `$ uname -srmo`: `Linux 5.15.0-79-generic x86_64 GNU/Linux`
- `$ gcc --version`: `gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0`
- `libwebsockets` built with the next cmake arguments (without encryption):
    ```shell
    cmake ../libwebsockets-main/ \
        -G "Unix Makefiles" \
        -D CMAKE_BUILD_TYPE:STRING=Release \
        -D CMAKE_INSTALL_PREFIX:PATH=`pwd`/../libwebsockets_installed/ \
        \
        -DLWS_WITH_SSL=OFF
    ```

**Comment by lws-team** [comment]
It's not crazy that there might be something to do here, but I could not casually reproduce it.

This is enough to break it on your toolchain?  On current Rocky 9 (gcc 11.3.1-4) there's no problem told.

/tmp/p.c:
```
#include <libwebsockets.h>
int main() { return 0; }
```

```
$ gcc -std=c99 -pedantic -pedantic-errors -Wall -Werror /tmp/p.c
$
```

**Comment by inobelar** [comment]
@lws-team I runned your command-line, but sligthly modified (added include & lib search pathes):
```
$ gcc -std=c99 -pedantic -pedantic-errors -Wall -Werror -I ./libwebsockets/installed/include -L ./libwebsockets/installed/lib/ -lwebsockets ./test_src.c -o test_app
```

And **confirm** same compilation errors, that I described in the issue. Perhaps these are not all errors, but only the first ones, after which the `gcc` refused to continue. Unfortunately, I don’t know how to quickly fix them in order to make PR.

Its not too critical to me, I can live without `pedantic` in my library, I just would like everything to be strict :)

**Comment by lws-team** [comment]
> I just would like everything to be strict :)

This is definitely the right way... I'm asking because I have to see it go wrong to know if I fixed it.  I'll try your exact toolchain on Ubuntu later now I know this test should be enough to make the problem appear.

**Comment by inobelar** [comment]
@lws-team Also, I found the same error messages, while searching on GitHub:
- Search: [redefinition of typedef 'lws_display_state_t'](https://github.com/search?q=redefinition+of+typedef+%27lws_display_state_t%27&type=issues)
  - https://github.com/warmcat/libwebsockets/issues/2739
  - https://github.com/open62541/open62541/issues/5171
- Search: [redefinition of typedef 'lhp_ctx_t'](https://github.com/search?q=redefinition+of+typedef+%27lhp_ctx_t%27&type=issues)
  - No additional issues found

**Comment by lws-team** [comment]
Yeah I wouldn't bother with random googling.

I installed 22.04 and 23.04, neither has the same gcc version as you mention... 22.04 has

gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0

and 23.04 has

gcc (Ubuntu 12.2.0-17ubuntu1) 12.2.0

Where did this toolchain actually come from?

**Comment by inobelar** [comment]
@lws-team Well, I'm using `Kubuntu 22.04`, and `lsb_release -a` show the next:
```
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04.3 LTS
Release:	22.04
Codename:	jammy
```

**Comment by lws-team** [comment]
Using my same test stub I showed before, there are no errors coming on 

$ gcc -std=c99 -pedantic -pedantic-errors -Wall -Werror /tmp/p.c

on either Ubuntu 22.04 (gcc 11.2.0-19) or 23.04 (gcc 12.2.0-17).  I'll take a look at Kubuntu tomorrow but it looks like it's pretty rare to see this symptom (or my test somehow doesn't trigger it).

**Comment by lws-team** [comment]
After realizing kubuntu is just ubuntu with specific packages, I installed the updated gcc pieces from Ubuntu repo and now have the same version

```
gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
```

But the test shown in https://github.com/warmcat/libwebsockets/issues/3002#issuecomment-1799118706 does not create any errors, nor when I use your version of the commandline with the extra -I.

I'm using current main lws.  What am I missing?

**Comment by inobelar** [comment]
@lws-team Good morning! I checked the compilation with `pedantic` on several Linux machines that were at hand (with different architectures, `gcc`/`clang` compilers, and toolchain versions) - the same commands were executed everywhere, the `libwebsockets` library was compiled successfully everywhere (because, library itself not compiled with `pedantic` mode), but include of its 'header' it in 'pedantic' context produces the same compilation errors (on all machines). On every machine `libwebsockets` was not installed 'from repository' or 'globally', only local build exists.

Below I also provide the reproduction steps that I literally followed on each machine. In short:
1. Clone the latest version of libwebsockets
2. Build without encryption & install locally (to not pollute global scope)
3. Create test sources file
4. Try to compile with include from there.

To me this looks like "reproducible on any compiler", and certainly not "quite rare". I'm surprised 🤯  that this is reproduced everywhere for me, but not for you 😅 

<details>
<summary>List of tested machines</summary>

- Codename: `Workstation` (This machine was mentioned previously, issue was started on it)
    - `lsb_release -a`:
        ```
        No LSB modules are available.
        Distributor ID: Ubuntu
        Description:    Ubuntu 22.04.3 LTS
        Release:        22.04
        Codename:       jammy
        ```
    - `uname -srmo`: `Linux 5.15.0-79-generic x86_64 GNU/Linux`
    - `gcc --version`: `gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0`
    - `clang --version`: `clang version 10.0.0-4ubuntu1 Target: x86_64-pc-linux-gnu`

- Codename: `Windows :: WSL`
    - `lsb_release -a`:
        ```
        No LSB modules are available.
        Distributor ID: Ubuntu
        Description:    Ubuntu 20.04.6 LTS
        Release:        20.04
        Codename:       focal
        ```
    - `uname -srmo`: `Linux 5.15.90.1-microsoft-standard-WSL2 x86_64 GNU/Linux`
    - `gcc --version`: `gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0`
    - `clang --version`: `clang version 10.0.0-4ubuntu1 Target: x86_64-pc-linux-gnu`

- Codename: `Rock Pi` (ROCK 3 Model B)
    - `lsb_release -a`:
        ```
        No LSB modules are available.
        Distributor ID: Debian
        Description:    Debian GNU/Linux 11 (bullseye)
        Release:        11
        Codename:       bullseye
        ```
    - `uname -srmo`: `Linux 4.4.194-11-rk3399-rockchip-g1bb08d49cc40 aarch64 GNU/Linux`
    - `gcc --version`: `gcc (Debian 10.2.1-6) 10.2.1 20210110`
    - `clang --version`: `Debian clang version 11.0.1-2 Target: aarch64-unknown-linux-gnu`

- Codename: `I.MX6` (Yep, here is pretty old toolchain)
    - `lsb_release -a`:
        ```
        No LSB modules are available.
        Distributor ID: Debian
        Description:    Debian GNU/Linux 8.9 (jessie)
        Release:        8.9
        Codename:       jessie
        ```
    - `uname -srmo`: `Linux 4.1.15-g14db131-dirty armv7l GNU/Linux`
    - `gcc --version`: `gcc (Debian 4.9.2-10+deb8u1) 4.9.2`
    - `clang --version`: `Debian clang version 3.5.0-10 (tags/RELEASE_350/final) (based on LLVM 3.5.0) Target: arm-unknown-linux-gnueabihf`

</details>

----

1. Clone `libwebsockets`, create `build` directory and go into it:
    ```shell
    $ git clone https://github.com/warmcat/libwebsockets
    $ mkdir build
    $ cd build
    ```

2. (Now we in `build` directory) - build library, install into `../installed/` directory (to not pollute global scope), and go out of `build` directory:
    ```shell
    $ cmake ../libwebsockets -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=`pwd`/../installed/ -DLWS_WITH_SSL=OFF
    $ make -j4
    $ make install
    $ cd -
    ```

----

Now we have the next directories structure:

```
.
├── build/
│   ├── CMakeCache.txt
│   ├── CMakeFiles/
│   ├── CPackConfig.cmake
│   ├── CPackSourceConfig.cmake
│   └── ... Rest of CMake build files ...
│
├── installed/
│   ├── bin/
│   │   ├── libwebsockets-test-client
│   │   ├── libwebsockets-test-lejp
│   │   ├── libwebsockets-test-server
│   │   └── libwebsockets-test-server-extpoll
│   ├── include/
│   │   ├── libwebsockets
│   │   │   ├── lws-adopt.h
│   │   │   ├── lws-async-dns.h
│   │   │   ├── lws-backtrace.h
│   │   │   ├── lws-bb-i2c.h
│   │   │   ├── lws-bb-spi.h
│   │   │   └── ... Rest of libwebsockets public headers ...
│   ├── lib/
│   │   ├── cmake/
│   │   │   └── libwebsockets/
│   │   │       ├── LibwebsocketsTargets-release.cmake
│   │   │       ├── LibwebsocketsTargets.cmake
│   │   │       ├── LwsCheckRequirements.cmake
│   │   │       ├── libwebsockets-config-version.cmake
│   │   │       └── libwebsockets-config.cmake
│   │   ├── libwebsockets.a
│   │   ├── libwebsockets.so -> libwebsockets.so.19
│   │   ├── libwebsockets.so.19
│   │   └── pkgconfig/
│   │       ├── libwebsockets.pc
│   │       └── libwebsockets_static.pc
│   └── share/
│       └── libwebsockets-test-server/
│           ├── candide-uncompressed.zip
│           ├── candide.zip
│           ├── favicon.ico
│           └── ... Rest of test-server files ...
│
└── libwebsockets/
    ├── CMakeLists-implied-options.txt
    ├── CMakeLists.txt
    ├── Kconfig
    ├── LICENSE
    ├── Makefile.projbuild
    ├── README.md
    ├── READMEs/
    ├── bug_report.md
    └── ... Rest of libwebsockets repo files ...
```

3. Create pretty-simple `test_src.c`:
    ```shell
    $ printf "#include <libwebsockets.h>\nint main() { return 0; }\n" > test_src.c
    ```

4. Try to compile `test_src.c`:

    - `gcc`:
        ```shell
        $ gcc -std=c99 -pedantic -pedantic-errors -Wall -Werror -I ./installed/include -L ./installed/lib/ -lwebsockets ./test_src.c -o test_app
        ```

        <details>
        <summary>gcc error messages</summary>
          
        ```
        In file included from ./installed/include/libwebsockets.h:786,
                         from ./test_src.c:1:
        ./installed/include/libwebsockets/lws-dlo.h:270:34: error: redefinition of typedef ‘lws_display_state_t’ [-Wpedantic]
          270 | typedef struct lws_display_state lws_display_state_t;
              |                                  ^~~~~~~~~~~~~~~~~~~
        In file included from ./installed/include/libwebsockets.h:785,
                         from ./test_src.c:1:
        ./installed/include/libwebsockets/lws-display.h:162:3: note: previous declaration of ‘lws_display_state_t’ was here
          162 | } lws_display_state_t;
              |   ^~~~~~~~~~~~~~~~~~~
        In file included from ./installed/include/libwebsockets.h:797,
                         from ./test_src.c:1:
        ./installed/include/libwebsockets/lws-html.h:588:3: error: redefinition of typedef ‘lhp_ctx_t’ [-Wpedantic]
          588 | } lhp_ctx_t;
              |   ^~~~~~~~~
        In file included from ./installed/include/libwebsockets.h:786,
                         from ./test_src.c:1:
        ./installed/include/libwebsockets/lws-dlo.h:463:24: note: previous declaration of ‘lhp_ctx_t’ was here
          463 | typedef struct lhp_ctx lhp_ctx_t;
              |                        ^~~~~~~~~
        ```
          
        </details>
    
    - `clang`:
        ```
        $ clang -std=c99 -pedantic -pedantic-errors -Wall -Werror -I ./installed/include -L ./installed/lib/ -lwebsockets ./test_src.c -o test_app
        ```

        <details>
        <summary>clang error messages</summary>
      
        ```
        In file included from ./test_src.c:1:
        In file included from ./installed/include/libwebsockets.h:786:
        ./installed/include/libwebsockets/lws-dlo.h:270:34: error: redefinition of typedef 'lws_display_state_t' is a C11 feature [-Werror,-Wtypedef-redefinition]
        typedef struct lws_display_state lws_display_state_t;
                                         ^
        ./installed/include/libwebsockets/lws-display.h:162:3: note: previous definition is here
        } lws_display_state_t;
          ^
        In file included from ./test_src.c:1:
        In file included from ./installed/include/libwebsockets.h:797:
        ./installed/include/libwebsockets/lws-html.h:588:3: error: redefinition of typedef 'lhp_ctx_t' is a C11 feature [-Werror,-Wtypedef-redefinition]
        } lhp_ctx_t;
          ^
        ./installed/include/libwebsockets/lws-dlo.h:463:24: note: previous definition is here
        typedef struct lhp_ctx lhp_ctx_t;
                               ^
        2 errors generated.
        ```
          
        </details>

Thanks for Your time!

**Comment by lws-team** [comment]
After struggling around with it I realized building from ./build, your -I and -L paths that I copy-pasted are wrong, they should be .. to use the installed ones from there.  Now I can finally join in with the cool guys that can get errors out -pedantic.

```
In file included from ../installed/include/libwebsockets.h:786,
                 from ./test_src.c:1:
../installed/include/libwebsockets/lws-dlo.h:270:34: warning: redefinition of typedef ‘lws_display_state_t’ [-Wpedantic]
  270 | typedef struct lws_display_state lws_display_state_t;
      |                                  ^~~~~~~~~~~~~~~~~~~
In file included from ../installed/include/libwebsockets.h:785,
                 from ./test_src.c:1:
../installed/include/libwebsockets/lws-display.h:162:3: note: previous declaration of ‘lws_display_state_t’ with type ‘lws_display_state_t’ {aka ‘struct lws_display_state’}
  162 | } lws_display_state_t;
      |   ^~~~~~~~~~~~~~~~~~~
In file included from ../installed/include/libwebsockets.h:797,
                 from ./test_src.c:1:
../installed/include/libwebsockets/lws-html.h:588:3: warning: redefinition of typedef ‘lhp_ctx_t’ [-Wpedantic]
  588 | } lhp_ctx_t;
      |   ^~~~~~~~~
In file included from ../installed/include/libwebsockets.h:786,
                 from ./test_src.c:1:
../installed/include/libwebsockets/lws-dlo.h:463:24: note: previous declaration of ‘lhp_ctx_t’ with type ‘lhp_ctx_t’ {aka ‘struct lhp_ctx’}
  463 | typedef struct lhp_ctx lhp_ctx_t;
      |                        ^~~~~~~~~
```

It's only those errors on the Rocky 9 gcc which is encouraging.

**Comment by lws-team** [comment]
That should be solved by 48e09ddf51ea014a60d666f8eb780cd801c0d4b5 on main now

**Comment by inobelar** [comment]
@lws-team thanks a lot! I tested that commit - everything everywhere works like a charm :)

Probably (for cleanup) you need also close this similar PR https://github.com/warmcat/libwebsockets/pull/2924
